ddl/ddl_api.go

```diff
@@ -221,18 +262,69 @@ func (d *ddl) ModifySchemaSetTiFlashReplica(ctx sessionctx.Context, stmt *ast.Al
 	if total == 0 {
 		return infoschema.ErrEmptyDatabase.GenWithStack("Empty database '%v'", dbName.O)
 	}
-	err := checkTiFlashReplicaCount(ctx, tiflashReplica.Count)
+	err := checkTiFlashReplicaCount(sctx, tiflashReplica.Count)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
+	var originVersion int64 = 0
+	var pendingCount uint32 = 0
+	forceCheck := false
+
+	logutil.BgLogger().Info("start batch add TiFlash replicas", zap.Int("total", total), zap.Int64("schemaID", dbInfo.ID))
 	for _, tbl := range dbInfo.Tables {
+		select {
+		case <-ctx.Done():
+			logutil.BgLogger().Info("abort batch add TiFlash replica", zap.Int64("schemaID", dbInfo.ID))
+			return nil
+		default:
+		}
+
 		tbReplicaInfo := tbl.TiFlashReplica
 		if !shouldModifyTiFlashReplica(tbReplicaInfo, tiflashReplica) {
 			logutil.BgLogger().Info("skip processing schema table", zap.Int64("tableID", tbl.ID), zap.Int64("schemaID", dbInfo.ID), zap.String("tableName", tbl.Name.String()), zap.String("schemaName", dbInfo.Name.String()))
 			skip += 1
 			continue
 		}
+
+		configRetry := tiflashCheckPendingTablesRetry
+		configWaitTime := tiflashCheckPendingTablesWaitTime
+		failpoint.Inject("FastFailCheckTiFlashPendingTables", func(value failpoint.Value) {
+			configRetry = value.(int)
+			configWaitTime = time.Second
+		})
+		// When handled some tables, we need to wait some tables become available.
+		if (succ+fail)%tiflashCheckPendingTablesTick == 0 || forceCheck {
+			// Maybe current schema is not up-to-date, so we should one ddlJob to update current schema from time to time.
+			forceCheck = true
+			pendingFunc := func() bool {
+				for retry := 0; retry < configRetry; retry += 1 {
+					d.getPendingTiFlashTableCount(&originVersion, &pendingCount)
+					threshold := getBatchPendingTiFlashCount(sctx)
+					delay := time.Duration(0)
+					if pendingCount >= threshold {
+						logutil.BgLogger().Info("too many unavailable tables, wait", zap.Uint32("threshold", threshold), zap.Uint32("current", pendingCount), zap.Int64("tableID", tbl.ID), zap.Duration("time", configWaitTime))
+						delay = configWaitTime
+					} else {
+						// If there are not many unavailable tables, we will no longer need force check.
+						forceCheck = false
+						return false
+					}
+					select {
+					case <-ctx.Done():
+						return true
+					case <-time.After(delay):
+					}
```

Comment:
Should select `context.Done()` first, maybe move code after branch `case <-time.After(delay):`.

Reply:
Splited into 2 selects

Reply:
Then it is even more complicated. Maybe consider sleep manually after `if pendingCount >= threshold {`. Then these code could just be the default branch of select.

Reply:
There are cases when ctx is done, and when this routine is sleeping.

Reply:
It doesn't matter to check it twice in a loop. Check in the end will follow a check at beginning soon.

Sleeping for a while is acceptable for exiting.

Reply:
Won't it cause goroutine leak?

Reply:
Won't it cause goroutine leak?

Reply:
If it eventually exits, then there is no leak.